File: scripts/nqcrc.lua
#!/usr/bin/env luajit

-- given an input file, outputs the NQCRC in hex

local nqcrcTable = {}
for i = 0, 255 do
	local value = bit.lshift(i, 0x18)
	for j = 0, 7 do
		local mod = 0
		if value < 0 then
			mod = 0xAF
		end
		value = bit.bxor(bit.lshift(value, 1), mod)
	end
	nqcrcTable[i] = value
end

local f = io.open(..., "rb")

local hash = 0
while true do
	local bt = f:read(1)
	if not bt then break end
	hash = bit.bxor(bit.lshift(hash, 8), nqcrcTable[bit.band(bit.bxor(bit.rshift(hash, 24), bt:byte()), 0xFF)])
end

if hash < 0 then
	hash = hash + 0x100000000
end
print(string.format("%08x", hash))

File: after/src/main/java/after/algorithms/NQCRC.java
package after.algorithms;

/**
 * Implementation of NQCRC.
 */
public class NQCRC {
	private static final int[] nqcrcTable = new int[256];
	static {
		for (int i = 0; i < nqcrcTable.length; i++) {
			int value = i << 0x18;
			for (int j = 0; j < 8; j++)
				value = (value << 1) ^ (value < 0 ? 0xAF : 0);
			nqcrcTable[i] = value;
		}
	}
	
	/**
	 * Advances NQCRC by a single byte.
	 */
	public static int hash(int hash, byte byt) {
		return (hash << 8) ^ nqcrcTable[(hash >>> 24) ^ (byt & 0xFF)];
	}

	/**
	 * Advances NQCRC by an array of bytes.
	 */
	public static int hash(int hash, byte[] data) {
		for (byte b : data)
			hash = hash(hash, b);
		return hash;
	}
}
